### DISCUSSION TITLE
How to repair UNTRACKS FILES

### DISCUSSION THREAD

**Discussion by AlliotTech** [immich-app/immich#4592] [Q&A]
Hi ,there is an question about the UNTRACKS FILES:  
<img width="1164" alt="image" src="https://github.com/immich-app/immich/assets/24980252/b2d57493-b5c5-42b2-8f4e-e22209156e61">
I found a lot of UNTRACKS FILES on the repair page, but the repair all button is not available. How should I fix them?
Thank you

**Comment by jrasm91** [comment]
Basically you should manually investigate and then re-upload or delete.

More likely than not, these look to be in your upload and may to partially uploaded files. You can manually inspect them, and if incomplete/corrupt you can delete them. If they are whole files that seem to be originals, I would check if you can find another version in the library already and if not manually move them out of that folder and re-upload them through the web UI. 

If there are untracked files in thumbs or encoded videos, you can more than likely delete them, as they are probably generated content that Immich did not clean up for some reason (past bug).

**Comment by AlliotTech** [comment]
I tried to delete some photos in web UI, and then remove it in trash, I found those photo PATH in "Repair" page. 
Is this a bug?

**Reply by jrasm91** [reply]
Yes that sounds like a bug that may or may not have been fixed already. The report is hopefully going to help us find those types of bugs, since it is easier to see them now.

**Comment by AlliotTech** [comment]
After reviewed one by one, I found these untracked files coming from:
- Videos or photos deleted from web UI and deleted from recycle bin
- Encoding unfinished videos or photos

The first one is most likely a bug. For the second case, can automatic deletion logic be added?

**Reply by AlliotTech** [reply]
By the way, server version : [v1.82.1](https://github.com/immich-app/immich/releases)

**Reply by briandipalma** [reply]
I think I'm seeing photos deleted from the web GUI also show up here. The files are still on the hard drive.

**Reply by briandipalma** [reply]
I grabbed the list from Immich, used a neovim macro to change the paths to moves and moved them out of the Immich library. They are all pictures I deleted but Immich didn't remove from the file system. Including their webp thumbnails it seems.

**Reply by briandipalma** [reply]
OK, that's all the "UNTRACKS FILES" fixed for me :) Including the thumbnails. Not ideal but maybe there was an issue with my server going down and the web GUI still letting me delete files without verifying the files were deleted or the new thrash functionality not deleting files that you delete from it?

**Reply by jrasm91** [reply]
We don't remove files from disk if they are marked as read only. Not sure if that applies here or not. Can you reproduce this with a brand new file on the latest release?

**Reply by AlliotTech** [reply]
But I am absolutely sure i didnt mark those photos read-only..

**Reply by jrasm91** [reply]
Ok. We can track this in the new issue, which I think has already been opened.

**Reply by briandipalma** [reply]
In my case yes it appears to be some sort of issue around ansible starting the docker compose job. If I start it manually it's fine but if ansible starts it it's not OK. Which is odd as I thought ansible was running as root. It looks like I'm wrong about that and I need to update the ansible script. Didn't realize non root users could run docker compose.

**Reply by AlliotTech** [reply]
Nope..I didn't use ansible to manage this immich docker container. I just manage it manualy.

**Reply by dakabali** [reply]
Hi all

I've got over 8k untracked files, all of them in external libraries that have been moved and rebinded again. These files exist and show up in the albums exacly as they should. Now that everything's great, could anyone provide a safe method (even a database SQL query) how this "untracked" status could be vanished? Thanks a lot.

<img width="1160" alt="kép" src="https://github.com/immich-app/immich/assets/12606599/f74c5d43-5bd0-478c-9572-dbee79d016b8">

**Reply by dakabali** [reply]
Okay I figured it out myself. It seemed to make sense to put the external (shared) library under the immich library because this could represent an in my understanding logical structure: user1 , user2, user3, ... , shared. The problem is that the library under the upload directory is being managed by immich and any manual interaction causes error messages. Getting my external library out of the immich library solved the issue imediately.

**Comment by GetGankNow** [comment]
Any update on this? I have 131 " UNTRACKS FILES "

**Reply by purpose-pending** [reply]
I envy you. 
<img width="196" alt="image" src="https://github.com/immich-app/immich/assets/40537713/b87df17b-ed80-4b4c-ac2c-91c47d9b4d40">

**Reply by tilianb** [reply]
Currently dealing with this issue on my server (11k+ untracked) and looking to using Immich CLI to re-upload.
I have moved out all the untracked files and uploading again using CLI. Currently to run on the server hosting Immich. Will provide an update shortly.

**Reply by nate-moo** [reply]
I have 17,744 lol, my immich instance crashed when emptying my trash after clearing out my library and I can't find a way to just delete those since they aren't supposed to be around still. These files are taking up nearly 75gb of space and I don't have the time to delete each one individually
![Screenshot_20240722-104819_Firefox Nightly](https://github.com/user-attachments/assets/5e32eda4-558c-4727-bdb3-3b92688228eb)

**Comment by tilianb** [comment]
Hi all, below are my instructions to fix the issue when there is a large number of untracked files. I have approx 12k+ files.

Log into Immich as admin > go to administration > repair
Review the list and manually move untracked files on your system to a temporary location (temp_location/). In my case this was all of my pictures after I made some changes to immich.

Install Immich CLI on the host or system you plan to use: https://immich.app/docs/features/command-line-interface

Log into your Immich instance (in my case running on the host I used localhost):
`immich login-key http://localhost:2283/api YOUR_API_KEY`

Upload all your files again (remove `--skip-hash` if you are unsure of duplicates):
`immich upload --skip-hash --recursive temp_location/`

Log into Immich as admin > go to administration > repair and review the list. You should be able to match up thumbnails and encoded videos again automatically through here.

If you have any issues or need clarification let me know

**Reply by Kqzz** [reply]
Thank you for this fix! Very slow, since I have a fairly large library, but working nonetheless.

**Reply by mekaneck** [reply]
Is there a programmatic way to move all the untracked files to a temporary location? I have 40,000+ "untracks" so I have no intention of doing that manually. I would like to find a programmatic way to delete the thumbs and video encodes, and then move the remaining files to a temporary location so they can be re-uploaded.

**Reply by jwillikers** [reply]
Here's the [Nushell](https://www.nushell.sh/) snippet I used to obtain a list of these files using the API, copy these files out of my server container to the current directory, and then delete them from the container.

```
http get --headers [Accept application/json x-api-key YOUR_ADMIN_API_KEY] https://immich.example.com/api/report | get extras | par-each {|it| ^podman cp $"immich-server-container:($it)" .; ^podman exec -it immich-server-container rm $it }
```

Alternatively, I'm sure it is possible to use the API to perform the download and deletion as well.

**Reply by ajitesh-kuppa-sivakumar** [reply]
Thanks @crazytil for the fix. 

If someone feels its slow, remember the jobs starts as soon as the media starts getting copied. I paused all the running job until the transfer is complete and repaired the untracked files. 

Hope it helps.

**Reply by joelis10** [reply]
do you have any idea how to do this for multiple users? it seems to work for me, the admin, but not for my other users for some reason. ```--skip-hash``` doesn't seem to do anything as it still checks their photos and skips duplicates

**Comment by ns-mglaske** [comment]
It seems that for me, most of my untracked files are in /usr/src/app/upload/thumbs, which seems like something I shouldn't mess with.  It's unclear if the thumbnail generator will managed these?

**Reply by jrasm91** [reply]
There have been some bugs in the past that leave thumbnails behind. I think external libraries might even have a current bug that leaves these behind. I'd probably just leave them and we'll probably add a way to automatically clean them up in the future.

**Reply by Kqzz** [reply]
This is the same for me.

I got this issue after reinstalling docker, and with it accidentally resetting my immich data. I'm not sure how this happened, but I lost the accounts information, but maintained the image library.

I'm not sure exactly how this happened, but if some way to recover from these files being untracked (but in the same structure as prior to this) that would be extremely helpful.

**Reply by essys** [reply]
@jrasm91 would this be the right ticket to keep track of the progress?
https://github.com/immich-app/immich/issues/4863

**Reply by Godd67** [reply]
So , is it a consensus of not doing anything with untracked files? I have ~75k untracked thumbnails from | /usr/src/app/upload/thumbs folder. Do they eat space?

**Comment by meminens** [comment]
<img width="1027" alt="image" src="https://github.com/immich-app/immich/assets/42714627/d0be2f5f-6c4d-419a-913a-ff4cfe2d2580">

I have untracked files in the `/usr/src/app/upload/upload` directory. How can I fix them? Should I follow @crazytil instructions?

**Reply by tilianb** [reply]
That's strange, are you able to confirm your upload directory?

**Reply by meminens** [reply]
`/home/misaligar/Docker/AppData/immich/library`

**Reply by j-seixas** [reply]
Same happened to me. Have 1 file in this situation. Opening shows correcly in portrait for like 1s, before changing to landscape (it's supposed to be landscape), then rotates to correct position and goes like this:

![image](https://github.com/immich-app/immich/assets/19241121/4f013410-86c3-483b-94e7-3f01f5977dc1)

It reports to be only 10.9MB (when original is 41.1MB). Checked the photo on immich web and it works great, and shows the correct size. Manually downloaded it and shows correctly (and size is the correct one). Can I just remove this one weird upload?

**Comment by kct-networx** [comment]
in my case, all untracked files were in Archive.
Repair reported 821 untracked files, but only 270 in Archive.
Unarchiving them fixed it for me.
I first tried to unarchive 10 Elements and that reduced the number of untracked files by ~30.

Strange.

That problem appeared in my case after an Upgrade from 1.94.1 to 1.95.1.

**Reply by 1P1E10** [reply]
Same here

**Comment by dnlldl** [comment]
Never went to the repair tab before, just noticed I have nearly 3000 untracked files, thanks guys :D !

Mostly thumbs, some uploads and some library.

**Reply by wildleaf** [reply]
For me they are all in thumbs folder. I guess it is leftovers from what I removed from external libraries. I just export them to untracked.txt and run a script inside the container against it to remove all the offending files.
![image](https://github.com/immich-app/immich/assets/1013099/ba21bb84-80c2-464e-9e3d-ac600ffd4a93)

**Reply by sym0nd0** [reply]
Are you able to share your script please? I have the same issue.

**Reply by SitramSoft** [reply]
Press the export button from the repair page and save the exported file with the path of each untracked file to `untracked.txt`. Rename the downloaded file to `untracked.sh` and give it execution permission `chmod +x untracked.sh`

Open the file with Sublime or whatever editor you prefer and make the following changes:

- Add `#!/bin/bash` and a new line at the beginning of the file

- Search for `/usr/src/app/` and replace it with `sudo rm -r <PATH_TO_YOUR_LIBRARY>`. Make sure to replace `<PATH_TO_YOUR_LIBRARY>` with the path to your library.

I ran the delete command for first file, then pressed `Refresh` in `Repair` page and I saw the file was no longer reported as untracked. Afterwards I executed the entire script `./utracked.sh` then pressed `Refresh` again in the `Repair` page and all  previously untracked files disappeared.

**Reply by sym0nd0** [reply]
Amazing, thank you.

**Reply by phil9309** [reply]
Is there a reason this isn't made available in the GUI yet? The Repair button is unclickable for me.

and just for clarification, it's supposed to look like this?
```

#!/bin/bash

MYPATH/upload/thumbs/29eb27e6-48d1-48a0-b565-7f047ed7d8af/38/b8/38b8b9ca-3cf2-48e0-b70c-deed4e510d85-thumbnail.webp

MYPATH/upload/thumbs/29eb27e6-48d1-48a0-b565-7f047ed7d8af/44/b2/44b2f543-b820-4f2b-993b-c84f15449e1b-thumbnail.webp

MYPATH/upload/thumbs/29eb27e6-48d1-48a0-b565-7f047ed7d8af/7c/21/7c210166-0e3b-45c8-9bee-8d20760e5897-thumbnail.webp
```

etc.... ??

**Reply by dnlldl** [reply]
> Is there a reason this isn't made available in the GUI yet? The Repair button is unclickable for me.
> 
> and just for clarification, it's supposed to look like this?
> 
> ```
> 
> #!/bin/bash
> 
> MYPATH/upload/thumbs/29eb27e6-48d1-48a0-b565-7f047ed7d8af/38/b8/38b8b9ca-3cf2-48e0-b70c-deed4e510d85-thumbnail.webp
> 
> MYPATH/upload/thumbs/29eb27e6-48d1-48a0-b565-7f047ed7d8af/44/b2/44b2f543-b820-4f2b-993b-c84f15449e1b-thumbnail.webp
> 
> MYPATH/upload/thumbs/29eb27e6-48d1-48a0-b565-7f047ed7d8af/7c/21/7c210166-0e3b-45c8-9bee-8d20760e5897-thumbnail.webp
> ```
> 
> etc.... ??

You are missing the `rm` part before. I highly suggest to NOT do this if you don't understand what you are doing. The Linux operating system is very unforgiving, it gives all the power to the user. With great power comes great responsability.

> Press the export button from the repair page and save the exported file with the path of each untracked file to `untracked.txt`. Rename the downloaded file to `untracked.sh` and give it execution permission `chmod +x untracked.sh`
> 
> Open the file with Sublime or whatever editor you prefer and make the following changes:
> 
> * Add `#!/bin/bash` and a new line at the beginning of the file
> * Search for `/usr/src/app/` and replace it with `sudo rm -r <PATH_TO_YOUR_LIBRARY>`. Make sure to replace `<PATH_TO_YOUR_LIBRARY>` with the path to your library.
> 
> I ran the delete command for first file, then pressed `Refresh` in `Repair` page and I saw the file was no longer reported as untracked. Afterwards I executed the entire script `./utracked.sh` then pressed `Refresh` again in the `Repair` page and all previously untracked files disappeared.

Please note that the `-r` option isn't required if you are going delete the files individually as you seem to be suggesting (like if you have 3000 files, you have 3000 separate `rm` commands). `sudo` will depend on your host as well. If you execute the command straight from the docker, you are already root and don't need to edit any path. You can use `docker exec -it <docker-name> bash` or similar to get a shell within the docker. 

For my specific case 99% of the files were thumbnails. I just flushed them out as I didn't see any problem regarding thumbnails on my side. I then ran the "Generate Thumbnails" job for missing thumbnails (none were missing so it did nothing).

**Reply by SitramSoft** [reply]
> Please note that the `-r` option isn't required if you are going delete the files individually as you seem to be suggesting (like if you have 3000 files, you have 3000 separate `rm` commands). `sudo` will depend on your host as well. If you execute the command straight from the docker, you are already root and don't need to edit any path. You can use `docker exec -it <docker-name> bash` or similar to get a shell within the docker.
> 
 You are correct. I usually add `-r` from force of habit so in case I delete a folder it recursively goes through it and delete it's content.

Again you are correct, I used`sudo` because the owner of the files and folders in immich library is `root` and I ran the commands from my linux pc not from the docker instance.

My example was meant as an indication on how to approach the job of manually clearing the untracked files, not as a general how-to. If you think it's confusing, I can remove the comment.

**Reply by dnlldl** [reply]
> > Please note that the `-r` option isn't required if you are going delete the files individually as you seem to be suggesting (like if you have 3000 files, you have 3000 separate `rm` commands). `sudo` will depend on your host as well. If you execute the command straight from the docker, you are already root and don't need to edit any path. You can use `docker exec -it <docker-name> bash` or similar to get a shell within the docker.
> 
> You are correct. I usually add `-r` from force of habit so in case I delete a folder it recursively goes through it and delete it's content.
> 
> Again you are correct, I used`sudo` because the owner of the files and folders in immich library is `root` and I ran the commands from my linux pc not from the docker instance.
> 
> My example was meant as an indication on how to approach the job of manually clearing the untracked files, not as a general how-to. If you think it's confusing, I can remove the comment.

I think it's relevant, I just want to stress out how people who have no clue what's going on should avoid toying with that. Anyone with decent experience with the Linux CLI should understand what's going on and be fine.

**Comment by jum0n** [comment]
I found over 40 of my files are untracked and they seem to be uploads that failed.  Since these are from my IOS phone, I'm not sure how best to check/re-upload since I dont have access to the file paths on the phone itself through the app.  I tried browsing it through USB but the folder structure is not even remotely similar and searching for the file name doesnt yield anything either.  I was hoping Immich was a solution for my family photo storage/backup needs but without even a basic function thats been in rsync and similar forever, I dont see how it could ever be trusted.

**Reply by essys** [reply]
I am sorry that you have had such a bad experience, which is most likely due to Apple's poor support of background synchronisation for any third-party apps, which is also a big of an issue for Nextcloud for example.

Oh and I hope you read the banner on Immich's home page, if not, here's a reminder:
<img width="1085" alt="Screenshot 2024-04-02 at 08 33 55" src="https://github.com/immich-app/immich/assets/7117619/cc60114d-1c3c-47ff-8408-e61f33d3ccac">

**Reply by jum0n** [reply]
Yea, I really love its capabilities otherwise.  I'm probably going to keep it around and see what develops and am truly grateful for what it does today, its far more than I need in many other respects.

**Reply by Qhilm** [reply]
For what it's worth, I have also a lot of untracked files (400+) as I just noticed but so far I didn't notice any missing file. Could it be the untracked files are failed uploads that the app retried later on and then successfully completed?

**Reply by jum0n** [reply]
Since the photos are never "moved" from the upload location, I dont think so (at least in my case).  There is no duplicate file name, it just seems to stop considering that file.  I tried removing one, and nothing replaces it on subsequent backups even if the source location is de-selected and re-selected on the client side.

**Comment by uninvitedguest** [comment]
The disk where I stored Immich uploads failed. The database is intact, and lucky me had backups of the upload folder. However, before restoring any files and while I was troubleshooting the failed disk, I accidentally let the Immich stack run.

As a result of Immich running while the disk was not mounted, Docker, in all its wisdom, recreated the missing path and continued to upload photos from my phone. I caught this and turned off uploads from my mobile, but when checking Immich's web app, I noticed that it was showing thumbnails for missing media from the failed disk.

![image](https://github.com/user-attachments/assets/d0cd8cdf-a41c-497f-a186-7863e27d36ca)

"No big deal," I thought, "as I'll restore those images from my backup and it will recognize them again."

Not so! After restoring the backed up upload folder those same blurred thumbnails exist, and the files are not being recognized despite being in their expected path.

Checking the repair page shows 103k untracked files. Seeing other discussions here, the solution to large amounts of untracked files has been to reupload them through the CLI - but this seems like a last resort solution to me as I would have to effectively start from scratch and lose out on all the albums that exist (many of which were ported over from Google Photos. Losing big points here with the wife if this happens). Is there a solution to give the database a kick and say "hey, the files you're looking for are right where you expect to find them!"

**Reply by bo0tzz** [reply]
If the files are where they are expected to be then Immich will just pick them right up. Double check your paths etc to make sure everything is as it should be.

**Reply by phil9309** [reply]
It depends on your folder structure I guess, you need to restore all of these, whereever they are for you.

`- ${EXTERNAL_PATH}:/usr/src/app/external
      - ${LIBRARY_LOCATION}:/usr/src/app/upload/library
      - ${UPLOAD_LOCATION}:/usr/src/app/upload/uploads
      - ${THUMBS_LOCATION}:/usr/src/app/upload/thumbs
      - ${PROFILE_LOCATION}:/usr/src/app/upload/profile
      - ${VIDEO_LOCATION}:/usr/src/app/upload/encoded-video`

**Comment by sannidhyaroy** [comment]
I use [this](https://github.com/clumsyCoder00/Immich-Tools/blob/main/move_untracked_files.sh) script for managing untracked files. I would recommend not to directly run the script, but dry-run it by editing the line containing the `rsync` command to include a `--dry-run` option. This way you can simulate the script and not directly proceed to modify anything before ensuring everything works correctly. Configure the script variables and make sure everything is working correctly. When you're confident, remove the `--dry-run` option from the line you edited before, and run the script.

**Comment by 1298345095** [comment]
How do I know which files on my iphone correspond to untracked files? Than I could replace it manually.

**Comment by sirweazel** [comment]
I read through some of this and i didn't see if there were options of other ways to handle this being discussed as a future feature?  If not, could maybe a couple options be implemented via the GUI?  Option 1: Button that creates an "untracked" album, and just moves the untracked files to the new album to be dealt with?  Option 2 with red flags and password confirmation and a warning: I'm adventures, and delete all untracked items?

**Comment by sirweazel** [comment]
I didn't realize this discussion was closed when posting.  Was there a decision around this topic, or was it tabled for later?  Sorry if i missed it in the previous posts.
